### ISSUE TITLE
[dev] Publishing documentation to readthedocs.org (from Github Actions?)

### ISSUE THREAD

**Issue by pramodk** [BlueBrain/CoreNeuron#460]
**Motivation** : Multiple projects like NEURON, CoreNEURON and NMODL builds documentation [as part of Travis](https://github.com/BlueBrain/CoreNeuron/blob/d443f9a839d0b2f2a0c5732f59db65400ec27fa3/.travis.yml#L182) or [Github Actions](https://github.com/BlueBrain/CoreNeuron/blob/d443f9a839d0b2f2a0c5732f59db65400ec27fa3/.github/workflows/coreneuron-ci.yml#L148) (very recently). This documentation is pushed to GitHub.io (e.g. NEURON doc [here](https://neuronsimulator.github.io/nrn/),  NMODL doc [here](https://bluebrain.github.io/nmodl/html/index.html); I am [seeing an error](http://bluebrain.github.io/bluebrain/coreneuron/index.html) for CoreNEURON somehow). 

One of the drawback of GitHub.io is that we can store only single snapshot at a time (without hack of branches). More elegant way is to use readthedocs.org where versioning and other aspects are managed very nicely! 

So it will be helpful if we could investigate readthedocs.org option and successfully migrate one of the project. 

**Action Items**:

* Check possibility of connecting Github Actions with readthedocs.org
* IIRC, readthedocs.org has an option to configure project where one can build the documentation on the readthedocs.org itself. Investigate what are options and best approaches.
* **NOTE** : there is one caveat I remember with above approach: for some projects setting up readthedocs.org configuration and building project like NMODL is not easy : 1. needed quite some build dependencies 2. need to build entire project to build full docs 3. also need to execute Jupyter notebooks (see [here](https://github.com/BlueBrain/nmodl/blob/d583ae80943b576f2498d1736dc7b0712c26fa0d/.travis.yml#L123)).  This is quite cumbersome ~2 years ago. And thats why if we can use GitHub Actions setup, that would be less hassle (?)

cc: @tristan0x  @matz-e @alkino @alexsavulescu

**Comment by alkino** [comment]
Publish sphinx and doxygen doc or only sphinx?

**Comment by tristan0x** [comment]
* Hosting the documentation on readthedocs is the preferred solution according to the BlueBrain software development guidelines.
* +1 to provide documentation for several versions.
* +1 to let readthedocs builds the documentation because having an external entity to build it ensures it can be done seamlessly. It also removes the burden of maintaining the build automaton ourself.

**Comment by matz-e** [comment]
> * +1 to let readthedocs builds the documentation because having an external entity to build it ensures it can be done seamlessly. It also removes the burden of maintaining the build automaton ourself.

I like this, although I'm curious about the implications of relying on breathe/exhale/all this doxygen-sphinx glue?

**Comment by alkino** [comment]
> > * +1 to let readthedocs builds the documentation because having an external entity to build it ensures it can be done seamlessly. It also removes the burden of maintaining the build automaton ourself.
> 
> I like this, although I'm curious about the implications of relying on breathe/exhale/all this doxygen-sphinx glue?

There is no doxygen-sphinx glue here, because we compile both separately and there is only a link from the sphinx documentation to the doxygen documentation.

**Comment by alexsavulescu** [comment]
Yes, there is no doxygen integration. CoreNEURON documentation is rather new and lacks substance. While breathe+exhale sound nice on paper, they are just extremely slow with NEURON.

> I am [seeing an error](http://bluebrain.github.io/bluebrain/coreneuron/index.html) for CoreNEURON somehow).

@pramodk  URI is case sensitive: http://bluebrain.github.io/CoreNeuron/

**Comment by pramodk** [comment]
> +1 to let readthedocs builds the documentation because having an external entity to build it ensures it can be done seamlessly. It also removes the burden of maintaining the build automaton ourself.

(This was discussed bit in #coreneuron Slack channel. Here is short summary)

@tristan0x : About building project on readthedocs: 

I mentioned **the caveat** in the description. When we were setting up documentation for NMODL project, almost ~2 years ago, I couldn't easily setup all build dependencies like flex, bison, cmake etc (https://github.com/readthedocs/readthedocs-docker-images/issues/114).

We needed to build nmodl python library in order to build documentation + compile Jupyter notebooks. Read-the-docs default workflow was not providing that much flexibility and hence we decided to use GitHub.io.

Looking at newer issues (https://github.com/readthedocs/readthedocs.org/issues/6919, https://github.com/readthedocs/readthedocs.org/issues/5387#issuecomment-468973915), there seems to be two ways install dependencies : 1) conf.py (hack) 2) conda config yaml

If above things work without too much trouble (or efforts), we can use above approach. Otherwise, we can think about simplified approaches.
